Support home channel change

MartinBroers opened the pull request:
Commit to support WIFI_EVENT_HOME_CHANNEL_CHANGE, introduced in https://github.com/espressif/esp-idf/commit/1427e14b724292ee8ceede6523d98cdee7c06037. I have opted to 'just' support the event and not propagate the values which come into the struct in the C library, but I could add those if required.

Also, I guarded it with a config directive (`#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]`, so not to break backwards builds. Maybe this is too restrictive, and we need to adjust for future builds (anything greater than 5.2), so  I am open to suggestions here. Or is the idea the next release works for 5.2 and not for older versions?

Also, I took the liberty to fix a few compilation errors and added 5.2 to the github workflow, but I am not sure this all is correct.

ivmarkov commented on src/bt/a2dp.rs:
```diff
@@ -134,6 +134,13 @@ pub enum ConnectionStatus {
     SuspendedByRemote = esp_a2d_audio_state_t_ESP_A2D_AUDIO_STATE_REMOTE_SUSPEND,
+    #[cfg(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
```
Nit: simplify this by replacing with `all(esp_idf_version_major = "5", any(esp_idf_version_minor = "0", esp_idf_version_minor = "1"))`

ivmarkov commented on src/bt/gap.rs:
```diff
@@ -3,6 +3,13 @@
 use core::convert::TryInto;
+#[cfg(any(
+    esp_idf_version_major = "4",
+    all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
```
Ditto.

ivmarkov commented on src/eth.rs:
```diff
@@ -462,6 +462,8 @@ where
                     spi_host_id: host,
                     spi_devcfg: &spi_devcfg as *const _ as *mut _,
                     int_gpio_num: int,
+                    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
This is not OK, as we'll have to touch it again once 5.3 or 6.0 is out. Rather, replace it with `not(<the upper condition>)`.

ivmarkov commented on src/eth.rs:
```diff
@@ -490,6 +492,8 @@ where
                     spi_host_id: host,
                     spi_devcfg: &spi_devcfg as *const _ as *mut _,
                     int_gpio_num: int,
+                    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto.

ivmarkov commented on src/eth.rs:
```diff
@@ -518,6 +522,8 @@ where
                     spi_host_id: host,
                     spi_devcfg: &spi_devcfg as *const _ as *mut _,
                     int_gpio_num: int,
+                    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto.

ivmarkov commented on src/wifi.rs:
```diff
@@ -1888,67 +1924,65 @@ impl<'a> EspEventDeserializer for WifiEvent<'a> {
+            wifi_event_t_WIFI_EVENT_ACTION_TX_STATUS => WifiEvent::ActionTxStatus,
+            wifi_event_t_WIFI_EVENT_STA_BEACON_TIMEOUT => WifiEvent::StaBeaconTimeout,
+            wifi_event_t_WIFI_EVENT_ROC_DONE => WifiEvent::RocDone,
+            #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto - see above.

ivmarkov commented on src/wifi.rs:
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
 #[repr(transparent)]
 pub struct WpsCredentialsRef(wifi_event_sta_wps_er_success_t__bindgen_ty_1);
 
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto - see above.

ivmarkov commented on src/wifi.rs:
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
+    new_snd: WifiSecondChan,
+}
+
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto - see above.

ivmarkov commented on src/wifi.rs:
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
+    Below,
+}
+
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto - see above.

ivmarkov commented on src/wifi.rs:
```diff
@@ -1873,6 +1906,9 @@ pub enum WifiEvent<'a> {
     ActionTxStatus,
     RocDone,
+
+    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Ditto - see above.

ivmarkov reviewed (changes requested):
I think it is in a very good shape. However, if you could fix the `#cfg`s - see my other comments.

MartinBroers commented:
Thanks @ivmarkov , some good suggestions.

MartinBroers commented:
Updated MR to 
- Compile again. Something in rust-nightly has changed causing clippy errors. Fixed those in the last commit
- Not use BT in the sample sdkconfig, so we don't need the update on esp-idf-sys.

ivmarkov commented on src/http/client.rs:
```diff
@@ -9,10 +9,9 @@ use core::cell::UnsafeCell;
 use alloc::collections::BTreeMap;
-use alloc::string::String;
-use alloc::string::ToString;
+#[cfg(not(feature = "default"))]
```
Should be `feature = "std"`, NOT `feature = "default"`.

ivmarkov commented on src/http/server.rs:
```diff
@@ -40,13 +40,11 @@ use core::time::*;
 extern crate alloc;
-use alloc::borrow::ToOwned;
-use alloc::boxed::Box;
+#[cfg(not(feature = "default"))]
```
Ditto.

ivmarkov commented on src/mdns.rs:
```diff
@@ -3,9 +3,8 @@
-use alloc::string::ToString;
-use alloc::vec::Vec;
+#[cfg(not(feature = "default"))]
+use alloc::{string::String, string::ToString, vec::Vec};
```
Ditto.

ivmarkov commented on src/mqtt/client.rs:
```diff
@@ -1,10 +1,10 @@
 use core::{slice, time};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Ditto

ivmarkov commented on src/timer.rs:
```diff
@@ -10,11 +10,11 @@
 use core::{ffi, ptr};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Ditto

ivmarkov commented on src/wifi.rs:
```diff
@@ -1,11 +1,11 @@
 use core::{cmp, ffi, fmt};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Ditto.

ivmarkov commented on src/ws/client.rs:
```diff
@@ -1,10 +1,13 @@
 
-use core::convert::{TryFrom, TryInto};
 use core::{ffi, time};
+#[cfg(not(feature = "default"))]
```
Ditto.

ivmarkov commented on .github/configs/sdkconfig.defaults:
```diff
@@ -18,17 +18,7 @@ CONFIG_ETH_SPI_ETHERNET_W5500=y
-CONFIG_BT_HFP_CLIENT_ENABLE=y
-CONFIG_BT_HFP_AUDIO_DATA_PATH_HCI=y
-CONFIG_BT_BLE_42_FEATURES_SUPPORTED=y
+CONFIG_BT_ENABLED=n
```
Please restore the BT settings, but commented out, and put a short comment on top: "Uncomment to enable Classic BT"

ivmarkov commented on src/espnow.rs:
```diff
@@ -10,6 +10,7 @@ use core::marker::PhantomData;
 
 use ::log::info;
 
+#[cfg(not(feature = "default"))]
```
See below.

ivmarkov commented on src/eth.rs:
```diff
@@ -6,6 +6,7 @@ use core::{ffi, ptr};
 use ::log::*;
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
See below.

ivmarkov commented on src/eventloop.rs:
```diff
@@ -1,13 +1,12 @@
 use core::{ffi, mem, ptr, slice};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
See below.

ivmarkov commented:
@MartinBroers W.r.t. the clippy errors... I'm thinking we need a more generic fix. As in - for example - not importing the `std` prelude - ever - even if the `std` feature is enabled. Otherwise we need to sprincle the whole code with conditional imports, which is not nice at all.
